src/services/datasources/services/datasourceRuns.js

```diff
@@ -105,6 +106,95 @@ class DatasourceRuns {
 		return datasourceRun;
 	}
 
+	/**
+	 * write a datasourceRun in pending state to the database
+	 * @param {Object} datasource a datasource object
+	 * @param {Bool} dryrun
+	 * @param {ObjectId} userId id of the user calling the run.
+	 */
+	async persistPendingRun(datasource, dryrun, userId) {
+		const dsrData = {
+			datasourceId: datasource._id,
+			status: PENDING,
+			config: datasource.config,
+			schoolId: datasource.schoolId,
+			dryrun,
+			createdBy: userId,
+		};
+		const dsr = await datasourceRunModel.create(dsrData);
+		return Promise.resolve(dsr);
+	}
+
+	/**
+	 * invokes a Syncer with params and data
+	 * @param {Object} params params for the syncer
+	 * @param {Object} data data for the syncer
+	 * @returns {Promise} SyncResult
+	 */
+	async startSync(params, data) {
+		// run a syncer
+		const promise = data
+			? this.app.service('sync').create(data, params)
+			: this.app.service('sync').find(params);
+		return promise;
+	}
+
+	/**
+	 * Update both the datasource and the datasourcerun with the results of a successful Sync.
+	 * @param {Object} result result of a syncer run
+	 * @param {Sting} logString string containing the log of the syncer run
+	 * @param {Date} startTime the time the sync has been started
+	 * @param {ObjectId} datasourceRunId id of the datasourceRun associated with the syncer run
+	 * @param {ObjectId} datasourceId id of the datasource
+	 */
+	async updateAfterSuccess(result, logString, startTime, datasourceRunId, datasourceId) {
+		const endTime = Date.now();
+		let status = SUCCESS;
+		result.forEach((e) => {
+			if (!e.success) status = ERROR;
+		});
+
+		// save to database
+		const updateData = {
+			status,
+			log: logString,
+			duration: endTime - startTime,
+		};
+
+		try {
+			await Promise.all([
+				datasourceRunModel.updateOne({ _id: datasourceRunId }, updateData),
+				this.app.service('datasources').patch(datasourceId, {
+					lastRun: endTime, lastStatus: status,
+				}),
+			]);
+		} catch (err) {
+			throw new GeneralError('error while updating datasourcerun', err);
+		}
+	}
+
+	/**
+	 * Update the datasource and datasourceRun with the results of a failed Sync
+	 * @param {String} message error message
+	 * @param {Date} startTime the time the sync has been started
+	 * @param {ObjectId} datasourceRunId id of the datasourceRun associated with the syncer run
+	 * @param {ObjectId} datasourceId id of the datasource
+	 */
+	async updateAfterFail(message, startTime, datasourceRunId, datasourceId) {
+		const endTime = Date.now();
+		const updateData = {
+			status: ERROR,
+			log: `Error while syncing: ${message}`,
+			duration: endTime - startTime,
+		};
+		await Promise.all([
+			datasourceRunModel.updateOne({ _id: datasourceRunId }, updateData),
+			this.app.service('datasources').patch(datasourceId, {
+				lastRun: endTime, lastStatus: ERROR,
+			}),
+		]);
```

Comment:
It exist no situation where it can fail, becouse the datasourceId and datasourceRunId is already validated?
Can the run, or source deleted, before it is finished?